cirq/contrib/quirk/cells/input_rotation_cells.py

```diff
@@ -0,0 +1,151 @@
+
+    def with_qubits(self, *new_qubits):
+        k = len(self.base_operation.qubits)
+        new_op_qubits = new_qubits[:k]
+        new_register = new_qubits[k:]
+        return QuirkInputRotationOperation(
+            self.identifier, new_register,
+            self.base_operation.with_qubits(*new_op_qubits), self.exponent_sign)
+
+    def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'):
+        sub_result = cirq.circuit_diagram_info(self.base_operation)
+        sign_char = '-' if self.exponent_sign == -1 else ''
+        symbols = list(sub_result.wire_symbols)
+        symbols.extend(f'A{i}' for i in range(len(self.register)))
+        qubit_index = (len(self.base_operation.controls) if isinstance(
+            self.base_operation, ops.ControlledOperation) else 0)
+        return cirq.CircuitDiagramInfo(
+            tuple(symbols),
+            exponent=f'({sign_char}A/2^{len(self.register)})',
+            exponent_qubit_index=qubit_index,
+            auto_exponent_parens=False)
+
+    def _has_unitary_(self):
+        return True
+
+    def _apply_unitary_(self, args: 'cirq.ApplyUnitaryArgs'):
+        transposed_args = args.with_axes_transposed_to_start()
+
+        target_axes = transposed_args.axes[:len(self.base_operation.qubits)]
+        control_axes = transposed_args.axes[len(self.base_operation.qubits):]
+        control_max = np.product([q.dimension for q in self.register]).item()
+
+        for i in range(control_max):
+            operation = self.base_operation**(self.exponent_sign * i /
+                                              control_max)
+            control_index = linalg.slice_for_qubits_equal_to(
+                control_axes, big_endian_qureg_value=i)
+            sub_args = cirq.ApplyUnitaryArgs(
+                transposed_args.target_tensor[control_index],
+                transposed_args.available_buffer[control_index], target_axes)
+            sub_result = cirq.apply_unitary(operation, sub_args)
+
+            if sub_result is not sub_args.target_tensor:
+                sub_args.target_tensor[...] = sub_result
+
+        return args.target_tensor
+
+    def __repr__(self):
+        return (f'cirq.contrib.quirk.QuirkInputRotationOperation('
+                f'identifier={self.identifier!r}, '
+                f'register={self.register!r}, '
+                f'base_operation={self.base_operation!r}, '
+                f'exponent_sign={self.exponent_sign!r})')
+
+
+def generate_all_input_rotation_cell_makers() -> Iterator[CellMaker]:
+    yield _input_rotation_gate("X^(A/2^n)", ops.X, +1)
+    yield _input_rotation_gate("Y^(A/2^n)", ops.Y, +1)
+    yield _input_rotation_gate("Z^(A/2^n)", ops.Z, +1)
+    yield _input_rotation_gate("X^(-A/2^n)", ops.X, -1)
+    yield _input_rotation_gate("Y^(-A/2^n)", ops.Y, -1)
+    yield _input_rotation_gate("Z^(-A/2^n)", ops.Z, -1)
+
+
+def _input_rotation_gate(identifier: str, gate: 'cirq.Gate',
+                         factor: int) -> CellMaker:
```

Comment:
nit: s/factor/exponent_sign/ for consistency

Reply:
Done.